### PR TITLE
Clock, Folder Update and Fixs 

### PR DESCRIPTION
**Pull request by VitoBarra**
implemented the clock chip that everyone wanted with fully support to Hz change

also added new feature regarding folders:
- renaming folder
- deleting folder
- moving chips between folders
- added a special folder to handle compatibility and chip without a valid folder 

Reworked the focus system used to handle interaction, now adding a new key-binding option should be easier 

A smal change to enhanced UX during editing a chip. now you can see the name the chip you are currently working on

issue #28, #34, #36 Fixed

issue #31 Fixed. it is a work around but it works pretty well so

issue #32 Fixed, this was an enormous debug work; in the end I found out the issue was caused by a single line of code: line 62 of the ChipPackage class. I commented the line and it looks like everything is still working properly.

issue #33 Fixed. I think I fixed this issue some time ago while I was working on updating compatibility. I ran some tests and it seems to be working for me however I think this needs further testing to assure it works for everyone.

### CODE REVIEW

**Comment by RobMayer** [comment]
hate to be that guy, but is there any word on if/when this will get folded in?

**Comment by crispeeweevile** [comment]
> hate to be that guy, but is there any word on if/when this will get folded in?

Dude, I don't blame you, I've been waiting for a while too. One thing to note is that there are still some bugs in this update, do they should be ironed out first, but you can totally download it and compile it to try it out yourself.

**Comment by RobMayer** [comment]
> > hate to be that guy, but is there any word on if/when this will get folded in?
> 
> Dude, I don't blame you, I've been waiting for a while too. One thing to note is that there are still some bugs in this update, do they should be ironed out first, but you can totally download it and compile it to try it out yourself.

totally could, and I may yet - I'm just not accustomed to unity in the slightest.

**Comment by crispeeweevile** [comment]
> > > hate to be that guy, but is there any word on if/when this will get folded in?
> > 
> > 
> > Dude, I don't blame you, I've been waiting for a while too. One thing to note is that there are still some bugs in this update, do they should be ironed out first, but you can totally download it and compile it to try it out yourself.
> 
> totally could, and I may yet - I'm just not accustomed to unity in the slightest.

Haha, yeah me neither, it's actually because of this project that I (very) recently started learning a little unity, thankfully just compiling this edition of the game wasn't too hard.

**Comment by AOx0** [comment]
> Haha, yeah me neither, it's actually because of this project that I (very) recently started learning a little unity, thankfully just compiling this edition of the game wasn't too hard.

Ahh, the beauty of open source. 

> hate to be that guy, but is there any word on if/when this will get folded in?

The thing is, I have a 128 GB laptop so downloading Unity and keeping it on my device becomes a nightmare when my average free storage is ~5GB, I have to literally uninstall every program I use on a daily basis until all things that I need to get done whit Unity are finished. So yeah, that's why almost all releases are on vacations or when I have a lot of spare time. Although, I will try to get this reviewed in the following week. Maybe it will get released on the weekend, if school doesn't get any more rude.

TL;DR I'll try to check it out next week. If everything is ok I will merge and release on the weekend, promise.

Edit: Thanks for showing interest to both of you. And thank you Vito for the PR, sorry for the very very very very late response :/

**Comment by AOx0** [comment]
Ups. Misclicked

**Comment by VitoBarra** [comment]
> > Haha, yeah me neither, it's actually because of this project that I (very) recently started learning a little unity, thankfully just compiling this edition of the game wasn't too hard.
> 
> Ahh, the beauty of open source.
> 
> > hate to be that guy, but is there any word on if/when this will get folded in?
> 
> The thing is, I have a 128 GB laptop so downloading Unity and keeping it on my device becomes a nightmare when my average free storage is ~5GB, I have to literally uninstall every program I use on a daily basis until all things that I need to get done whit Unity are finished. So yeah, that's why almost all releases are on vacations or when I have a lot of spare time. Although, I will try to get this reviewed in the following week. Maybe it will get released on the weekend, if school doesn't get any more rude.
> 
> TL;DR I'll try to check it out next week. If everything is ok I will merge and release on the weekend, promise.
> 
> Edit: Thanks for showing interest to both of you. And thank you Vito for the PR, sorry for the very very very very late response :/

haha that's no problem, I totally understand the time problem in fact I would like to do more for the project but university require time and so I cannot  :C

File: Assets/Scripts/Interaction/ChipInterfaceEditor.cs [hunk 1]
```diff
@@ -4,577 +4,608 @@
...
+            dragHandleStartY = (selectedSignals[indexA].transform.position.y +
+                                selectedSignals[indexB].transform.position.y) /
+                               2f;
+        }
+        else
+        {
+            dragHandleStartY = selectedSignals[selectedSignals.Count / 2].transform.position.y;
+        }
+
+        PropertiesMenu.EnableUI(this, selectedSignals[0].signalName, IsGroup, selectedSignals[0].useTwosComplement,
```
Comment (AOx0) [inline Assets/Scripts/Interaction/ChipInterfaceEditor.cs]: I belive `IsGroup` reports wrong values.

The problem with this is that the signed checkbox of pin groups is never shown.

Screenshot 2022-12-02 at 7 46 32 p m


Shouldn't this be called something like:

```suggestion
        PropertiesMenu.EnableUI(this, selectedSignals[0].signalName, selectedSignals.Count() > 1, selectedSignals[0].useTwosComplement,
```

File: Assets/Scripts/Interaction/ChipInterfaceEditor.cs [hunk 2]
```diff
@@ -4,577 +4,608 @@
...
+    bool isDragging;
+    float dragHandleStartY;
+    float dragMouseStartY;
+
+    // Grouping
+    int currentGroupSize = 1;
+    int currentGroupID;
+    Dictionary<int, ChipSignal[]> groupsByID;
+
+    private bool IsGroup => currentGroupSize > 1;
```
Comment (AOx0) [inline Assets/Scripts/Interaction/ChipInterfaceEditor.cs]: I believe the problem is here. It relies in `currentGroupSize`.

File: Assets/Scripts/Interaction/ChipInterfaceEditor.cs [hunk 3]
```diff
@@ -4,577 +4,608 @@
...
+                visiblePins.AddRange(spawnedSignal.inputPins);
+                visiblePins.AddRange(spawnedSignal.outputPins);
+                spawnedSignals[i] = spawnedSignal;
+            }
+
+            if (IsGroup)
+            {
+                groupsByID.Add(currentGroupID, spawnedSignals);
+                // Reset group size after spawning
+                currentGroupSize = 1;
```
Comment (AOx0) [inline Assets/Scripts/Interaction/ChipInterfaceEditor.cs:356]: But `currentGroupSize` changes it's value temporarily to handle the spawning of new groups and then resets to 1.

Thus, it is not a reliable source for getting the current selected group size.

**Review by AOx0 — APPROVED** [review]
It will get merged with PR #37
